cql3/expr/expression.cc

```diff
@@ -1659,6 +1659,53 @@ cql3::raw_value evaluate(const conjunction& conj, const evaluation_inputs& input
     return raw_value::make_value(boolean_type->decompose(true));
 }
 
+cql3::raw_value evaluate(const field_selection& field_select, const evaluation_inputs& inputs) {
+    const user_type_impl* udt_type = dynamic_cast<const user_type_impl*>(&field_select.type->without_reversed());
+    if (udt_type == nullptr) {
+        on_internal_error(expr_logger, "evaluate(field_selection): type is not a user defined type");
+    }
+
+    sstring selected_field_name = field_select.field->to_string();
+    std::optional<std::size_t> selected_field_idx = udt_type->idx_of_field(to_bytes(selected_field_name));
+    if (!selected_field_idx.has_value()) {
+        throw exceptions::invalid_request_exception(
+            format("Unkown field '{}' in expression {}, user type {} doesn't have a field with this name.",
+                   selected_field_name, field_select, udt_type->get_name_as_cql_string()));
+    }
+
+    cql3::raw_value udt_value = evaluate(field_select.structure, inputs);
+    if (udt_value.is_null()) {
+        return cql3::raw_value::make_null();
+    }
+
+    // Serialized UDT value has the same representation as a serialized tuple,
+    // so we can use read_nth_tuple_elem to read the nth UDT field.
+    std::optional<managed_bytes_opt> element_value = udt_value.view().with_value(
+        [&](const FragmentedView auto& value_bytes) -> std::optional<managed_bytes_opt> {
+            // std::optional<std::optional<FragmentedView>> field_read
```

Comment:
I didn't know how to write it when there's no clear way to get the type of `value_bytes`.
It's `const FragmentedView auto& value_bytes`, but how do I get what this `auto` is? `FragmentedView` is just a concept, not a type name.
I guess it's possible to acquire this type using some cool C++ code, but it was the quickest to just spell it out in the comment.

Reply:
I'm not familiar with "with_value()", doesn't it return a known type?
But if it doesn't, maybe something like
```
using value_type = std::remove_reference<decltype(value_bytes)>::type;
std::optional<std::optional<value_type>> ...
```
But that's not beautiful either... So you can just keep your version.

Reply:
> I'm not familiar with "with_value()", doesn't it return a known type?

`with_value` is a method that takes a closure which can accept argument of any type which fulfills the concept `FragmentedView`, and runs it on the `raw_value`. The argument is some view of the bytes, but we don't know which one, it could be fragmented or not.
The return type is known, but `value_bytes` has the `FragmentedView` type, which is different from the result type.
```c++
    template <typename Func>
    requires std::invocable<Func, const managed_bytes_view&> && std::invocable<Func, const fragmented_temporary_buffer::view&>
    decltype(auto) with_value(Func f) const {
        switch (_data.index()) {
        case 0: return f(std::get<fragmented_temporary_buffer::view>(_data));
        default: return f(std::get<managed_bytes_view>(_data));
        }
    }
```